dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java

```diff
@@ -0,0 +1,243 @@
+import com.ibm.ws.runtime.metadata.ComponentMetaData;
+import com.ibm.ws.runtime.metadata.MetaDataSlot;
+import com.ibm.ws.threadContext.ComponentMetaDataAccessorImpl;
+import com.ibm.wsspi.kernel.feature.LibertyFeature;
+
+import io.openliberty.checkpoint.spi.CheckpointHook;
+import io.openliberty.checkpoint.spi.CheckpointPhase;
+import io.openliberty.microprofile.telemetry.internal.common.constants.OpenTelemetryConstants;
+import io.openliberty.microprofile.telemetry.internal.interfaces.OpenTelemetryInfoFactory;
+
+@Component(service = { ApplicationStateListener.class, OpenTelemetryLifecycleManager.class }, property = { "service.vendor=IBM", "service.ranking:Integer=1500" })
+public class OpenTelemtryLifecycleManagerImpl implements ApplicationStateListener, OpenTelemetryLifecycleManager {
+
+    private static final TraceComponent tc = Tr.register(OpenTelemtryLifecycleManagerImpl.class);
+
+    private final MetaDataSlot slotForOpenTelemetryInfoHolder;
+    private final OpenTelemetryInfoFactory openTelemetryInfoFactory;
+
+    //These three are set during activation, and refreshed on checkpoint restore. (runtimeInstance is only set if we need one)
+    private boolean waitingForCheckpointRestore;
+    private boolean isRuntimeEnabled;
+    private LazyInitializer<OpenTelemetryInfo> runtimeInstance = null;
+
+    @Activate
+    public OpenTelemtryLifecycleManagerImpl(@Reference MetaDataSlotService slotService, @Reference OpenTelemetryInfoFactory openTelemetryInfoFactory,
+                                            @Reference(service = LibertyFeature.class, target = "(ibm.featureName=mpTelemetry-*)") ServiceReference<LibertyFeature> featureRef) {
+
+        this.openTelemetryInfoFactory = openTelemetryInfoFactory;
+        slotForOpenTelemetryInfoHolder = slotService.reserveMetaDataSlot(ApplicationMetaData.class);
+
+        String shortName = (String) featureRef.getProperty("ibm.featureName");
+        //the runtimeMode was added in 2.0
+        final boolean telemetry2OrLater = shortName.startsWith("mpTelemetry-1") ? false : true;
+
+        CheckpointHook checkpointHook = new CheckpointHook() {
+
+            @Override
+            public void restore() {
+                waitingForCheckpointRestore = false;
+                checkThenSetRuntimeFields(telemetry2OrLater);
+            }
+        };
+
+        // If checkpoint is enabled this statement will register the hook we created above. Then checkpoint returns true so the if statement is false.
+        // If checkpoint is disabled it does nothing and returns false, so the if statement will be true.
+        if (!CheckpointPhase.getPhase().addMultiThreadedHook(checkpointHook)) {
+            //Checkpoint disabled
+            waitingForCheckpointRestore = false;
+            checkThenSetRuntimeFields(telemetry2OrLater);
+        } else {
+            //Checkpoint enabled, before restore.
+            waitingForCheckpointRestore = true;
+        }
+    }
```

Comment:
Since you do the same thing in the checkpoint hook restore method and the case where the hook is not added this is a candidate for using `CheckpointPhase.onRestore`.  Something like this:

```java
waitingForCheckpointRestore = true;
CheckpointPhase.onRestore(() -> {
    waitingForCheckpointRestore = false;
    checkThenSetRuntimeFields(telemetry2OrLater);
}
```

The lambda will run synchronously if not doing a checkpoint or will run on restore for the checkpoint case.

This is just an improvement suggestion.  You can decide to do that later I think.  I'll leave it up to you.

Reply:
I like this a lot, it will make a good tool for my arsenal. Since this PR is a blocker for others I'll implement it in a follow up rather than get another build.